### PR TITLE
refactor: Parallelize Engine::MoveShip

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature**

This PR contains the first part of #10217.

## Summary
This PR pulls in the additional library needed for parallelization, and makes the necessary build changes. It also parallelizes `Ship::Move` as an "easy gains" change, as recommended by @petervdmeer; this also allows testing the new concurrency management code.

The `Ship::Move` change moves tractor beam and anti-missile counting out of `Ship::Move`, changes the ship list to a vector, and changes the visuals vector to a list.

Special thanks to Leaf for helping with the macos issues.

## Testing Done
I tested that the game does not crash, and that the thread sanitizer doesn't report any errors.

## Wiki Update
Not needed, as our build docs are kept in this repository.

## Performance Impact
With an omnis save (for reproducability) with 5k ships:

Before:
Idle CPU usage was around 160%, and `Ship::Move` was about 56.8% of CalculateStep.
image

[before.collapsed](https://github.com/user-attachments/files/16365484/EndlessSky_2024-07-24-191121.collapsed.txt)

After:
Idle CPU usage was around 90%, and `Ship::Move` was about 14.3% of CalculateStep.
image

[after.collapsed](https://github.com/user-attachments/files/16365405/EndlessSky_2024-07-24-185552.collapsed.txt)

### CODE REVIEW

File: source/Engine.cpp [hunk 1]
```diff
@@ -131,11 +151,36 @@ namespace {
...
+	void Prune(list<Type> &objects)
+	{
+		for(auto it = objects.begin(); it != objects.end(); )
+		{
+			if(it->ShouldBeRemoved())
+				it = objects.erase(it);
+			else
+				++it;
+		}
+	}
```
Comment (RisingLeaf) [inline source/Engine.cpp:164]: std::erase_if ?
https://en.cppreference.com/w/cpp/container/list/erase2

Comment (tibetiroka) [reply source/Engine.cpp:164]: That's the plan, but 
image

Comment (RisingLeaf) [reply source/Engine.cpp:164]: oh ouch

Comment (RisingLeaf) [reply source/Engine.cpp:164]: remove_if ?

Comment (tibetiroka) [reply source/Engine.cpp:165]: I'd rather just do the refactor once. C++20 is coming soon anyway.

**Comment by XessWaffle** [comment]
Just a quick question, as this might affect #10348, do your changes to Ship::Move include parallelizing the parts of the internal logic of the function, or is it that the function call itself is parallelized across multiple ships while keeping the logic the same?

**Comment by tibetiroka** [comment]
> do your changes to Ship::Move include parallelizing the parts of the internal logic of the function, or is it that the function call itself is parallelized across multiple ships while keeping the logic the same?

It parallelizes the entire function across multiple ships. If you don't use any Engine or AI-global variables, you should be fine. (And I can fix the rest fairly easily.)

File: .gitignore [hunk 2]
```diff
@@ -27,6 +27,7 @@ scons-local*
 .sconsign.dblite
 /*.d
 /vcpkg/
+tbb/
```
Comment (RisingLeaf) [inline .gitignore:30]: can this please go into a subfolder named external or something alike, I know this is a temporary thing, but I fear that other libs might do the same thing and it would get really crowded in my es folder?

File: source/AI.cpp [hunk 3]
```diff
@@ -362,7 +362,7 @@ namespace {
 
 
 
-AI::AI(const PlayerInfo &player, const List<Ship> &ships,
+AI::AI(const PlayerInfo &player, const std::vector<std::shared_ptr<Ship>> &ships,
```
Comment (RisingLeaf) [inline source/AI.cpp:365]:
```suggestion
AI::AI(const PlayerInfo &player, const vector<shared_ptr<Ship>> &ships,
```

File: source/AI.cpp [hunk 4]
```diff
@@ -1033,7 +1033,8 @@ void AI::Step(Command &activeCommands)
 				// Find the possible parents for orphaned fighters and drones.
 				auto parentChoices = vector<shared_ptr<Ship>>{};
 				parentChoices.reserve(ships.size() * .1);
-				auto getParentFrom = [&it, &gov, &parentChoices](const list<shared_ptr<Ship>> &otherShips) -> shared_ptr<Ship>
+				auto getParentFrom = [&it, &gov, &parentChoices](const std::vector<std::shared_ptr<Ship>> &otherShips)
```
Comment (RisingLeaf) [inline source/AI.cpp:1036]:
```suggestion
				auto getParentFrom = [&it, &gov, &parentChoices](const vector<shared_ptr<Ship>> &otherShips)
```

File: source/Engine.cpp [hunk 5]
```diff
@@ -100,6 +101,26 @@ namespace {
 		return Radar::UNFRIENDLY;
 	}
 
+	// TODO (C++20): Replace with erase_if.
```
Comment (RisingLeaf) [inline source/Engine.cpp:104]: this comment belongs to  the other prune i think

File: source/Engine.cpp [hunk 6]
```diff
@@ -131,11 +152,36 @@ namespace {
...
+		{
+			objects.swap(added);
+			objects.insert(objects.end(), make_move_iterator(added.begin()), make_move_iterator(added.end()));
+			added.clear();
+		}
+		else
+		{
+			objects.insert(objects.end(), make_move_iterator(added.begin()), make_move_iterator(added.end()));
+			added.clear();
+		}
```
Comment (RisingLeaf) [inline source/Engine.cpp:184]: this is an unrelated performance upgrade isnt it?

File: source/Engine.cpp [hunk 7]
```diff
@@ -1468,23 +1515,36 @@ void Engine::CalculateStep()
...
 		bool wasUntargetable = !it->IsTargetable();
-		MoveShip(it);
+		MoveShip(it, shipResourceProvider);
 		bool isTargetable = it->IsTargetable();
 		if(flagshipSystem == it->GetSystem()
 			&& ((wasUntargetable && isTargetable) || flagshipBecameTargetable)
 			&& isTargetable && flagshipIsTargetable)
 				eventQueue.emplace_back(player.FlagshipPtr(), it, ShipEvent::ENCOUNTER);
-	}
+	});
```
Comment (RisingLeaf) [inline source/Engine.cpp:1534]:
```suggestion
	});

```

File: source/concurrent/PartiallyGuarded.h [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+	void emplace_front(Args &&...args) = delete;
+
+
+
+private:
+	std::mutex write_mutex;
+};
+
+template<class T, class Allocator = std::allocator<T>>
+class PartiallyGuardedVector : public std::vector<T, Allocator> {
```
Comment (RisingLeaf) [inline source/concurrent/PartiallyGuarded.h:63]: one file per class or one class per file

File: source/concurrent/ResourceProvider.h [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LOCK_PROVIDER_H_
+#define LOCK_PROVIDER_H_
```
Comment (RisingLeaf) [inline source/concurrent/ResourceProvider.h:17]: RESOURCE_PROVIDER_H_

File: source/concurrent/ResourceProvider.h [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+private:
+	// Default-initialize the correct number of members in each per-thread resource.
+	template<size_t Index = 0>
+	typename std::enable_if_t<Index < sizeof...(Types), void>
+	inline initialize_vectors(size_t size);
+
+	template<size_t Index = 0>
+	typename std::enable_if_t<Index == sizeof...(Types), void>
+	inline initialize_vectors(size_t size);
+
```
Comment (RisingLeaf) [inline source/concurrent/ResourceProvider.h:113]:
```suggestion


```

File: source/Engine.h [hunk 11]
```diff
@@ -159,7 +161,13 @@ class Engine {
 
 	void CalculateStep();
 
-	void MoveShip(const std::shared_ptr<Ship> &ship);
+	// Thread-safe wrapper for MoveShip.
+	using ShipResourceProvider = ResourceProvider<std::list<Visual>, std::list<std::shared_ptr<Flotsam>>,
```
Comment (RisingLeaf) [inline source/Engine.h:165]: are you using this class ResourceProvider anywhere else? Do you plan to? If not, we can make this a much more simple object that does not need to be template mayhem

File: source/concurrent/WithMutex.h [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+#ifndef WITH_MUTEX_H_
+#define WITH_MUTEX_H_
+
+#include <mutex>
+
+
+
+// Helper class to add a mutex to Body without declaring the full copy constructor.
+// These operations are not thread safe. Do not copy the object in a concurrent context,
+// or with a locked mutex.
```
Comment (RisingLeaf) [inline source/concurrent/WithMutex.h:25]: Instead of having a comment that people likely wont read, this should be enforced by code

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/Engine.cpp (hunk 5 above)
Comment (tibetiroka) [reply source/Engine.cpp:104]: Can't we replace all of them?

File: source/Engine.cpp (hunk 6 above)
Comment (tibetiroka) [reply source/Engine.cpp:184]: Not entirely unrelated, but it's not essential to the PR.

Comment (RisingLeaf) [reply source/Engine.cpp:184]: I mean it is your decision, but stuff is faster to review if it is split up. but then again, it is just a small change

File: source/Engine.h (hunk 11 above)
Comment (tibetiroka) [reply source/Engine.h:165]: It's going to be used in more places. Visuals will get a provider and BatchDrawList will also have a provider.

Comment (RisingLeaf) [reply source/Engine.h:165]: in that case its good

File: source/Engine.cpp (hunk 5 above)
Comment (RisingLeaf) [reply source/Engine.cpp:104]: I am not quite sure, for some reason this one was made more complex, it already uses its own remove_if for some reason

File: source/concurrent/WithMutex.h (hunk 12 above)
Comment (tibetiroka) [reply source/concurrent/WithMutex.h:25]: If it could be enforced easily, it would be enforced in the standard library... There isn't really a "good" way to do it, just various trade-offs. This code works just fine unless you copy a body with a locked mutex, and expect the copied mutex to also be locked. But even if I lock the copy's mutex, the same comment would have to be made, as that wouldn't always be safe with temporaries...

File: .gitignore (hunk 2 above)
Comment (tibetiroka) [reply .gitignore:30]: Sure, I can rename it. Btw, if you have access to a windows machine, can you check if you can get it to build by extacting tbb to the same place where the mingw libraries are (and tweaking cmake config to account for it)? That would mean only the CI would have to use this workaround.

~~Well it's not any worse than having catch's source code in our tree~~

File: source/concurrent/ResourceProvider.h (hunk 9 above)
Comment (tibetiroka) [inline source/concurrent/ResourceProvider.h:17]:
```suggestion
#ifndef RESOURCE_PROVIDER_H_
#define RESOURCE_PROVIDER_H_
```

File: .gitignore (hunk 2 above)
Comment (RisingLeaf) [reply .gitignore:30]: I do not have a windows machine, sorry

File: source/CMakeLists.txt [hunk 13]
```diff
@@ -364,6 +364,13 @@ target_sources(EndlessSkyLib PRIVATE
 	comparators/ByGivenOrder.h
 	comparators/ByName.h
 	comparators/BySeriesAndIndex.h
+	concurrent/Parallel.h
+	concurrent/PartiallyGuardedList.h
+	concurrent/PartiallyGuardedMap.h
+	concurrent/PartiallyGuardedVector.h
```
Comment (RisingLeaf) [inline source/CMakeLists.txt:370]: Can these two be added when they are actually used?
PartiallyGuardedMap and PartiallyGuardedVector
would make more sense to me

**Comment by RisingLeaf** [comment]
One last thing that would interest me is memory usage, if you could give a bit of insight there that would be cool.

**Comment by tibetiroka** [comment]
To be honest, I haven't really checked that. In theory, the only changes on our side would be the extra buffered heap memory from the ResourceProvider's containers. If you have, say, 5k ships, and have a ResourceProvider managing a vector of ship pointers, that would cost around 64 bits * 5k = 40KB of extra memory, plus some minimal overhead for the ResourceProvider object itself.

TBB might have some extra memory overhead, and pulling in more STL functions does increase the executable's size (which does increase memory usage), but it shouldn't be significant.

**Comment by RisingLeaf** [comment]
do you know what type of threading tbb uses?
If you say there wont be a significant impact I believe you, just want to be on the safe side.

**Comment by tibetiroka** [comment]
It's mainly relying on fork-join pools: https://oneapi-src.github.io/oneTBB/main/tbb_userguide/How_Task_Scheduler_Works.html

**Comment by RisingLeaf** [comment]
You could add a (fetchcontent)[https://cmake.org/cmake/help/latest/module/FetchContent.html] to cmake to avoid the curl and unzipping of tbb

**Comment by RisingLeaf** [comment]
optional of course

**Comment by tibetiroka** [comment]
Huh that actually sounds like a much better way of doing this.

**Comment by RisingLeaf** [comment]
FetchContent is a very cool tool

File: source/Engine.cpp [hunk 14]
```diff
@@ -1695,9 +1756,19 @@ void Engine::CalculateStep()
 
 
 
+// Move a ship. Can be called concurrently.
```
Comment (RisingLeaf) [inline source/Engine.cpp:1759]:
```suggestion
// Wrapper for MoveShip() that can be called in a concurrent context.
```
or use the full actual comment:
```
// Move a ship. Also determine if the ship should generate hyperspace sounds or
// boarding events, fire weapons, and launch fighters.
```

File: source/Engine.h [hunk 15]
```diff
@@ -159,7 +161,13 @@ class Engine {
...
+			std::vector<std::shared_ptr<Ship>>, std::vector<Projectile>>;
+	void MoveShip(const std::shared_ptr<Ship> &ship, ShipResourceProvider &provider);
+	void MoveShip(const std::shared_ptr<Ship> &ship, std::list<Visual> &visuals,
```
Comment (RisingLeaf) [inline source/Engine.h:168]: ~~I think it was you who wanted me to synchronize comments between header and source~~

File: source/Planet.cpp [hunk 16]
```diff
@@ -689,26 +689,25 @@ string Planet::DemandTribute(PlayerInfo &player) const
...
 	Personality fighterPersonality = Personality::DefenderFighter();
-	for(auto it = defenders.begin(); it != end; ++it)
+	for(const auto &it : defenders)
 	{
-		(**it).SetPersonality(defenderPersonality);
-		if((**it).HasBays())
-			for(auto bay = (**it).Bays().begin(); bay != (**it).Bays().end(); ++bay)
+		it->SetPersonality(defenderPersonality);
+		if(it->HasBays())
+			for(auto bay = it->Bays().begin(); bay != it->Bays().end(); ++bay)
```
Comment (RisingLeaf) [inline source/Planet.cpp:710]: why no for(const auto& bay : it->Bays()) ?

File: source/concurrent/PartiallyGuardedList.h [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+#include <mutex>
+
+
+
+// Vectors that guard against most concurrent modifications. Use with caution.
+// For most purposes, this can be treated and passed around like any stl container,
+// but any function modifying it inside a concurrent context MUST receive it as a PartiallyGuarded object
+// to ensure proper handling.
+template<class T, class Allocator = std::allocator<T>>
+class PartiallyGuardedList : public std::list<T, Allocator> {
```
Comment (RisingLeaf) [inline source/concurrent/PartiallyGuardedList.h:29]: would it maybe be easier to have PartiallyGuardedList  not inherit from std::list, but store an instance of it?
that way we only have to provide the functions we need and do not have to delete everything else

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/concurrent/PartiallyGuardedList.h (hunk 17 above)
Comment (tibetiroka) [reply source/concurrent/PartiallyGuardedList.h:29]: No, because there are functions that I want to make available unguarded, and I also didn't want to touch any of the code where I pass these objects around.

File: source/Planet.cpp (hunk 16 above)
Comment (tibetiroka) [reply source/Planet.cpp:710]: That's not what the original code had, but I can change it if you prefer.

Comment (RisingLeaf) [reply source/Planet.cpp:710]: dew it

File: source/concurrent/PartiallyGuardedList.h (hunk 17 above)
Comment (RisingLeaf) [reply source/concurrent/PartiallyGuardedList.h:29]: > because there are functions that I want to make available unguarded,

would be possible with that approach

> and I also didn't want to touch any of the code where I pass these objects around.

ok

File: source/concurrent/ResourceProvider.h [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+	std::array<std::mutex, sizeof...(Types)> remoteLocks;
+	// Locks for each per-thread resource.
+	std::vector<std::mutex> locks;
+};
+
+
+
+template <class ...Types>
+ResourceProvider<Types...>::ResourceGuard::ResourceGuard(size_t index, ResourceProvider<Types...> &provider)
+		: index(index), provider(provider)
```
Comment (RisingLeaf) [inline source/concurrent/ResourceProvider.h:129]:
```suggestion
	: index(index), provider(provider)
```

File: source/concurrent/ResourceProvider.h [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+	size_t id = std::hash<std::thread::id>{}(std::this_thread::get_id());
+	const size_t index = id % Size();
+	if(locks[index].try_lock())
+		return ResourceGuard(index, *this);
+	else
+	{
+		for(unsigned newIndex = 0; newIndex < Size(); newIndex++)
+			if(locks[newIndex].try_lock())
+				return ResourceGuard(newIndex, *this);
+	}
```
Comment (RisingLeaf) [inline source/concurrent/ResourceProvider.h:290]:
```suggestion

	for(unsigned newIndex = 0; newIndex < Size(); newIndex++)
		if(locks[newIndex].try_lock())
			return ResourceGuard(newIndex, *this);

```

File: source/concurrent/PartiallyGuardedList.h [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+	void push_front(T &&value) = delete;
+	void insert(typename std::list<T, Allocator>::const_iterator pos, const T& value) = delete;
+	template<class InputIt>
+	void insert(typename std::list<T, Allocator>::const_iterator pos, InputIt first, InputIt last) = delete;
+	template<class ...Args>
+	void emplace(typename std::list<T, Allocator>::const_iterator pos, Args &&...args) = delete;
+	void pop_back() = delete;
+	void pop_front() = delete;
+	template<class ...Args>
+	void emplace_front(Args &&...args) = delete;
```
Comment (RisingLeaf) [inline source/concurrent/PartiallyGuardedList.h:52]: what about clear(), insert_range(), erase(), resize(), swap()?

**Review by RisingLeaf — CHANGES_REQUESTED** [review]

File: source/concurrent/PartiallyGuardedList.h (hunk 20 above)
Comment (RisingLeaf) [reply source/concurrent/PartiallyGuardedList.h:52]: forget insert_range() thats c++23

Comment (tibetiroka) [reply source/concurrent/PartiallyGuardedList.h:52]: Some of them we don't use, others are used only sequentially (and are thus unguarded)

Also, did you mean splice instead of swap?

Comment (RisingLeaf) [reply source/concurrent/PartiallyGuardedList.h:52]: no, I meant swap, but thanks for the explanation (maybe make a comment about which ones are unguarded?)

**Review by RisingLeaf — APPROVED** [review]

**Comment by RisingLeaf** [comment]
requested peter as he seemed to know a bit (or more idk) about concurrency from the old PR

File: source/concurrent/PartiallyGuardedList.h [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+// for concurrent modifications.
+template<class T, class Allocator = std::allocator<T>>
+class PartiallyGuardedList : public std::list<T, Allocator> {
+public:
+	template<class... Args>
+	inline T &emplace_back(Args &&... args)
+	{
+		const std::lock_guard<std::mutex> lock(write_mutex);
+		return std::list<T, Allocator>::emplace_back(args...);
+	}
```
Comment (RisingLeaf) [inline source/concurrent/PartiallyGuardedList.h:38]: ah just noticed this now, no can do, move downwards according to style guide

Comment (tibetiroka) [reply source/concurrent/PartiallyGuardedList.h:38]: I thought small inline functions were allowed here?

Comment (RisingLeaf) [reply source/concurrent/PartiallyGuardedList.h]: Nope

**Comment by tibetiroka** [comment]
I don't really have the time to maintain this PR right now. I might reopen this later if there is a serious interest for getting this into the game and I have more time.
